File: web/src/services/zquant/factor.ts
// Copyright 2025 ZQuant Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the Apache License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Author: kevin
// Contact:
//     - Email: kevin@vip.qq.com
//     - Wechat: zquant2025
//     - Issues: https://github.com/yoyoung/zquant/issues
//     - Documentation: https://github.com/yoyoung/zquant/blob/main/README.md
//     - Repository: https://github.com/yoyoung/zquant

import { request } from '@umijs/max';

/**
 * 获取因子定义列表
 * POST /api/v1/factor/definitions/query
 */
export async function getFactorDefinitions(params?: {
  skip?: number;
  limit?: number;
  enabled?: boolean;
  order_by?: string;
  order?: string;
}) {
  return request<ZQuant.FactorDefinitionListResponse>('/api/v1/factor/definitions/query', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: params || {},
  });
}

/**
 * 创建因子定义
 * POST /api/v1/factor/definitions
 */
export async function createFactorDefinition(body: ZQuant.FactorDefinitionCreate) {
  return request<ZQuant.FactorDefinitionResponse>('/api/v1/factor/definitions', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
  });
}

/**
 * 更新因子定义
 * POST /api/v1/factor/definitions/update
 */
export async function updateFactorDefinition(id: number, body: ZQuant.FactorDefinitionUpdate) {
  return request<ZQuant.FactorDefinitionResponse>('/api/v1/factor/definitions/update', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: { ...body, factor_id: id },
  });
}

/**
 * 删除因子定义
 * POST /api/v1/factor/definitions/delete
 */
export async function deleteFactorDefinition(id: number) {
  return request<void>('/api/v1/factor/definitions/delete', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: { factor_id: id },
  });
}

/**
 * 获取因子定义详情
 * POST /api/v1/factor/definitions/get
 */
export async function getFactorDefinition(id: number) {
  return request<ZQuant.FactorDefinitionResponse>('/api/v1/factor/definitions/get', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: { factor_id: id },
  });
}

/**
 * 获取因子配置列表
 * POST /api/v1/factor/configs/query
 */
export async function getFactorConfigs(params?: {
  skip?: number;
  limit?: number;
  enabled?: boolean;
  order_by?: string;
  order?: string;
}) {
  return request<ZQuant.FactorConfigListResponse>('/api/v1/factor/configs/query', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: params || {},
  });
}

/**
 * 获取单个因子配置
 * POST /api/v1/factor/configs/get
 */
export async function getFactorConfigById(factorId: number) {
  return request<ZQuant.FactorConfigResponse>('/api/v1/factor/configs/get', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: { factor_id: factorId },
  });
}

/**
 * 更新因子配置
 * POST /api/v1/factor/configs/update
 */
export async function updateFactorConfigById(
  factorId: number,
  body: {
    mappings?: Array<{
      model_id: number;
      codes: string[] | null;
    }> | null;
    enabled?: boolean | null;
  }
) {
  return request<ZQuant.FactorConfigResponse>('/api/v1/factor/configs/update', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: { ...body, factor_id: factorId },
  });
}

/**
 * 删除因子配置
 * POST /api/v1/factor/configs/delete
 */
export async function deleteFactorConfigById(factorId: number) {
  return request<void>('/api/v1/factor/configs/delete', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: { factor_id: factorId },
  });
}

/**
 * 获取因子配置（JSON格式）（已废弃，向后兼容）
 * POST /api/v1/factor/definitions/config
 */
export async function getFactorConfig(factorId: number) {
  return request<{
    enabled: boolean;
    mappings: Array<{
      model_id: number;
      codes: string[] | null;
    }>;
  }>('/api/v1/factor/definitions/config', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: { factor_id: factorId },
  });
}

/**
 * 更新因子配置（JSON格式）
 * POST /api/v1/factor/definitions/config/update
 */
export async function updateFactorConfigJson(
  factorId: number,
  body: {
    enabled: boolean;
    mappings: Array<{
      model_id: number;
      codes: string[] | null;
    }>;
  }
) {
  return request<{
    enabled: boolean;
    mappings: Array<{
      model_id: number;
      codes: string[] | null;
    }>;
  }>('/api/v1/factor/definitions/config/update', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: { ...body, factor_id: factorId },
  });
}

/**
 * 删除因子配置（清空配置）
 * POST /api/v1/factor/definitions/config/delete
 */
export async function deleteFactorConfigJson(factorId: number) {
  return request<void>('/api/v1/factor/definitions/config/delete', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: { factor_id: factorId },
  });
}

/**
 * 获取因子模型列表
 * POST /api/v1/factor/models/query
 */
export async function getFactorModels(params?: {
  factor_id?: number;
  skip?: number;
  limit?: number;
  enabled?: boolean;
  order_by?: string;
  order?: string;
}) {
  return request<ZQuant.FactorModelListResponse>('/api/v1/factor/models/query', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: params || {},
  });
}

/**
 * 创建因子模型
 * POST /api/v1/factor/models
 */
export async function createFactorModel(body: ZQuant.FactorModelCreate) {
  return request<ZQuant.FactorModelResponse>('/api/v1/factor/models', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
  });
}

/**
 * 更新因子模型
 * POST /api/v1/factor/models/update
 */
export async function updateFactorModel(id: number, body: ZQuant.FactorModelUpdate) {
  return request<ZQuant.FactorModelResponse>('/api/v1/factor/models/update', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: { ...body, model_id: id },
  });
}

/**
 * 删除因子模型
 * POST /api/v1/factor/models/delete
 */
export async function deleteFactorModel(id: number) {
  return request<void>('/api/v1/factor/models/delete', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: { model_id: id },
  });
}

/**
 * 获取因子模型详情
 * POST /api/v1/factor/models/get
 */
export async function getFactorModel(id: number) {
  return request<ZQuant.FactorModelResponse>('/api/v1/factor/models/get', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: { model_id: id },
  });
}

/**
 * 创建因子配置
 * POST /api/v1/factor/configs
 */
export async function createFactorConfig(body: ZQuant.FactorConfigCreate) {
  return request<ZQuant.FactorConfigResponse>('/api/v1/factor/configs', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
  });
}

/**
 * 获取因子配置列表（按因子分组）（已废弃）
 * POST /api/v1/factor/configs/grouped
 */
export async function getFactorConfigsGrouped(params?: {
  enabled?: boolean;
}) {
  return request<ZQuant.FactorConfigGroupedListResponse>('/api/v1/factor/configs/grouped', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: params || {},
  });
}

/**
 * 更新因子配置（按因子ID，支持多映射）
 * POST /api/v1/factor/configs/by-factor/update
 */
export async function updateFactorConfigByFactor(factorId: number, body: ZQuant.FactorConfigUpdate) {
  return request<ZQuant.FactorConfigGroupedResponse>('/api/v1/factor/configs/by-factor/update', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: { ...body, factor_id: factorId },
  });
}

/**
 * 更新单个因子配置
 * POST /api/v1/factor/configs/update_single
 */
export async function updateFactorConfig(id: number, body: ZQuant.FactorConfigSingleUpdate) {
  return request<ZQuant.FactorConfigResponse>('/api/v1/factor/configs/update_single', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: { ...body, config_id: id },
  });
}

/**
 * 删除因子配置
 * POST /api/v1/factor/configs/delete_single
 */
export async function deleteFactorConfig(id: number) {
  return request<void>('/api/v1/factor/configs/delete_single', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: { config_id: id },
  });
}

/**
 * 删除因子配置（按因子ID，删除该因子的所有配置）
 * POST /api/v1/factor/configs/by-factor/delete
 */
export async function deleteFactorConfigByFactor(factorId: number) {
  return request<void>('/api/v1/factor/configs/by-factor/delete', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: { factor_id: factorId },
  });
}

/**
 * 手动触发因子计算
 * POST /api/v1/factor/calculate
 */
export async function calculateFactor(body: ZQuant.FactorCalculationRequest) {
  return request<ZQuant.FactorCalculationResponse>('/api/v1/factor/calculate', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
  });
}

/**
 * 查询因子计算结果
 * POST /api/v1/factor/results
 */
export async function getFactorResults(body: ZQuant.FactorResultQueryRequest) {
  return request<ZQuant.FactorResultResponse>('/api/v1/factor/results', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
  });
}

/**
 * 查询量化因子数据
 * POST /api/v1/factor/quant-factors/query
 */
export async function getQuantFactors(params: {
  ts_code?: string;
  start_date?: string;
  end_date?: string;
  filter_conditions?: any;
  skip?: number;
  limit?: number;
  order_by?: string;
  order?: string;
}) {
  return request<ZQuant.QuantFactorQueryResponse>('/api/v1/factor/quant-factors/query', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: params,
  });
}

File: zquant/schemas/factor.py
# Copyright 2025 ZQuant Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the Apache License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# Author: kevin
# Contact:
#     - Email: kevin@vip.qq.com
#     - Wechat: zquant2025
#     - Issues: https://github.com/yoyoung/zquant/issues
#     - Documentation: https://github.com/yoyoung/zquant/blob/main/README.md
#     - Repository: https://github.com/yoyoung/zquant

"""
因子相关Pydantic模型
"""

from datetime import date, datetime
from typing import Any

from pydantic import BaseModel, Field

from zquant.schemas.common import QueryRequest


# ==================== 因子定义 Schema ====================

class FactorDefinitionCreate(BaseModel):
    """创建因子定义请求"""

    factor_name: str = Field(..., description="因子名称（唯一标识）")
    cn_name: str = Field(..., description="中文简称")
    en_name: str | None = Field(None, description="英文简称")
    column_name: str = Field(..., description="因子表数据列名")
    description: str | None = Field(None, description="因子详细描述")
    factor_type: str | None = Field("单因子", description="因子类型：单因子、组合因子")
    enabled: bool = Field(True, description="是否启用")
    factor_config: dict[str, Any] | None = Field(None, description="因子配置，格式：{\"enabled\": bool, \"mappings\": [{\"model_id\": int, \"codes\": list[str]|None}, ...]}")


class FactorDefinitionUpdate(BaseModel):
    """更新因子定义请求"""

    factor_id: int = Field(..., description="因子ID")
    cn_name: str | None = Field(None, description="中文简称")
    en_name: str | None = Field(None, description="英文简称")
    column_name: str | None = Field(None, description="因子表数据列名")
    description: str | None = Field(None, description="因子详细描述")
    factor_type: str | None = Field(None, description="因子类型：单因子、组合因子")
    enabled: bool | None = Field(None, description="是否启用")
    factor_config: dict[str, Any] | None = Field(None, description="因子配置")


class FactorDefinitionDeleteRequest(BaseModel):
    """删除因子定义请求模型"""
    factor_id: int = Field(..., description="因子ID")


class FactorDefinitionResponse(BaseModel):
    """因子定义响应"""

    id: int
    factor_name: str
    cn_name: str
    en_name: str | None
    column_name: str
    description: str | None
    factor_type: str | None
    enabled: bool
    factor_config: dict[str, Any] | None
    created_by: str | None
    created_time: datetime
    updated_by: str | None
    updated_time: datetime

    class Config:
        from_attributes = True

    @classmethod
    def from_orm(cls, obj):
        """从ORM对象创建响应"""
        data = {
            "id": obj.id,
            "factor_name": obj.factor_name,
            "cn_name": obj.cn_name,
            "en_name": obj.en_name,
            "column_name": obj.column_name,
            "description": obj.description,
            "factor_type": obj.factor_type,
            "enabled": obj.enabled,
            "factor_config": obj.get_factor_config(),
            "created_by": getattr(obj, "created_by", None),
            "created_time": obj.created_time,
            "updated_by": getattr(obj, "updated_by", None),
            "updated_time": obj.updated_time,
        }
        return cls(**data)


class FactorDefinitionListRequest(QueryRequest):
    """因子定义列表查询请求模型"""
    enabled: bool | None = Field(None, description="是否启用")
    order_by: str | None = Field(
        None, description="排序字段：id, factor_name, cn_name, created_time, updated_time"
    )


class FactorDefinitionGetRequest(BaseModel):
    """获取因子定义请求模型"""
    factor_id: int = Field(..., description="因子ID")


class FactorDefinitionListResponse(BaseModel):
    """因子定义列表响应"""

    items: list[FactorDefinitionResponse] = Field(..., description="因子定义列表")
    total: int = Field(..., description="总数")


# ==================== 因子模型 Schema ====================

class FactorModelCreate(BaseModel):
    """创建因子模型请求"""

    factor_id: int = Field(..., description="因子ID")
    model_name: str = Field(..., description="模型名称")
    model_code: str = Field(..., description="模型代码（用于识别计算器类型）")
    config_json: dict[str, Any] | None = Field(None, description="模型配置（JSON格式）")
    is_default: bool = Field(False, description="是否默认算法")
    enabled: bool = Field(True, description="是否启用")

    class Config:
        protected_namespaces = ()


class FactorModelUpdate(BaseModel):
    """更新因子模型请求"""

    model_id: int = Field(..., description="模型ID")
    model_name: str | None = Field(None, description="模型名称")
    model_code: str | None = Field(None, description="模型代码")
    config_json: dict[str, Any] | None = Field(None, description="模型配置")
    is_default: bool | None = Field(None, description="是否默认算法")
    enabled: bool | None = Field(None, description="是否启用")

    class Config:
        protected_namespaces = ()


class FactorModelDeleteRequest(BaseModel):
    """删除因子模型请求模型"""
    model_id: int = Field(..., description="模型ID")

    class Config:
        protected_namespaces = ()


class FactorModelResponse(BaseModel):
    """因子模型响应"""

    id: int
    factor_id: int
    model_name: str
    model_code: str
    config_json: dict[str, Any] | None
    is_default: bool
    enabled: bool
    created_by: str | None
    created_time: datetime
    updated_by: str | None
    updated_time: datetime

    class Config:
        from_attributes = True
        protected_namespaces = ()

    @classmethod
    def from_orm(cls, obj):
        """从ORM对象创建响应"""
        data = {
            "id": obj.id,
            "factor_id": obj.factor_id,
            "model_name": obj.model_name,
            "model_code": obj.model_code,
            "config_json": obj.get_config(),
            "is_default": obj.is_default,
            "enabled": obj.enabled,
            "created_by": getattr(obj, "created_by", None),
            "created_time": obj.created_time,
            "updated_by": getattr(obj, "updated_by", None),
            "updated_time": obj.updated_time,
        }
        return cls(**data)


class FactorModelListRequest(QueryRequest):
    """因子模型列表查询请求模型"""
    factor_id: int | None = Field(None, description="因子ID")
    enabled: bool | None = Field(None, description="是否启用")
    order_by: str | None = Field(
        None, description="排序字段：id, model_name, model_code, created_time, updated_time"
    )

    class Config:
        protected_namespaces = ()


class FactorModelGetRequest(BaseModel):
    """获取因子模型请求模型"""
    model_id: int = Field(..., description="模型ID")

    class Config:
        protected_namespaces = ()


class FactorModelListResponse(BaseModel):
    """因子模型列表响应"""

    items: list[FactorModelResponse] = Field(..., description="因子模型列表")
    total: int = Field(..., description="总数")


# ==================== 因子配置 Schema ====================

class FactorConfigMappingItem(BaseModel):
    """因子配置映射项（模型-代码列表对）"""

    model_id: int = Field(..., description="模型ID")
    codes: list[str] | None = Field(None, description="股票代码列表，None或空列表表示默认配置")

    class Config:
        protected_namespaces = ()


class FactorConfigCreate(BaseModel):
    """创建因子配置请求（支持多个映射）"""

    factor_id: int = Field(..., description="因子ID")
    mappings: list[FactorConfigMappingItem] = Field(..., description="模型-代码列表映射对列表")
    enabled: bool = Field(True, description="是否启用")
    
    def to_config_dict(self) -> dict[str, Any]:
        """转换为配置字典"""
        return {
            "enabled": self.enabled,
            "mappings": [{"model_id": m.model_id, "codes": m.codes} for m in self.mappings]
        }


class FactorConfigUpdate(BaseModel):
    """更新因子配置请求（支持批量更新映射）"""

    factor_id: int = Field(..., description="因子ID")
    mappings: list[FactorConfigMappingItem] | None = Field(None, description="模型-代码列表映射对列表（None表示不更新）")
    enabled: bool | None = Field(None, description="是否启用")


class FactorConfigDeleteRequest(BaseModel):
    """删除因子配置请求模型"""
    factor_id: int = Field(..., description="因子ID")


class FactorConfigSingleUpdate(BaseModel):
    """更新单个因子配置请求（向后兼容）"""

    config_id: int = Field(..., description="配置ID")
    model_id: int | None = Field(None, description="模型ID")
    codes: list[str] | None = Field(None, description="股票代码列表")
    enabled: bool | None = Field(None, description="是否启用")

    class Config:
        protected_namespaces = ()


class FactorConfigSingleDeleteRequest(BaseModel):
    """删除单个因子配置请求模型"""
    config_id: int = Field(..., description="配置ID")

    class Config:
        protected_namespaces = ()


class FactorConfigResponse(BaseModel):
    """因子配置响应"""

    factor_id: int = Field(..., description="因子ID（主键）")
    config: dict[str, Any] = Field(..., description="因子配置，格式：{\"enabled\": bool, \"mappings\": [{\"model_id\": int, \"codes\": list[str]|None}, ...]}")
    enabled: bool = Field(..., description="是否启用")
    created_by: str | None = Field(None, description="创建人")
    created_time: datetime = Field(..., description="创建时间")
    updated_by: str | None = Field(None, description="修改人")
    updated_time: datetime = Field(..., description="更新时间")

    class Config:
        from_attributes = True
        protected_namespaces = ()

    @classmethod
    def from_orm(cls, obj):
        """从ORM对象创建响应"""
        data = {
            "factor_id": obj.factor_id,
            "config": obj.get_config(),
            "enabled": obj.enabled,
            "created_by": getattr(obj, "created_by", None),
            "created_time": obj.created_time,
            "updated_by": getattr(obj, "updated_by", None),
            "updated_time": obj.updated_time,
        }
        return cls(**data)


class FactorConfigListRequest(QueryRequest):
    """因子配置列表查询请求模型"""
    enabled: bool | None = Field(None, description="是否启用")
    order_by: str | None = Field(None, description="排序字段：factor_id, created_time, updated_time")


class FactorConfigGetRequest(BaseModel):
    """获取因子配置请求模型"""
    factor_id: int = Field(..., description="因子ID")


class FactorConfigGroupedListRequest(BaseModel):
    """因子配置分组列表查询请求模型"""
    enabled: bool | None = Field(None, description="是否启用")


class FactorConfigListResponse(BaseModel):
    """因子配置列表响应"""

    items: list[FactorConfigResponse] = Field(..., description="因子配置列表")
    total: int = Field(..., description="总数")


class FactorConfigGroupedResponse(BaseModel):
    """因子配置分组响应（按factor_id分组）"""

    factor_id: int = Field(..., description="因子ID")
    enabled: bool = Field(..., description="是否启用")
    mappings: list[FactorConfigResponse] = Field(..., description="该因子的所有配置映射")
    created_by: str | None = Field(None, description="创建人")
    created_time: datetime = Field(..., description="创建时间")
    updated_by: str | None = Field(None, description="修改人")
    updated_time: datetime = Field(..., description="更新时间")


class FactorConfigGroupedListResponse(BaseModel):
    """因子配置分组列表响应"""

    items: list[FactorConfigGroupedResponse] = Field(..., description="因子配置分组列表")
    total: int = Field(..., description="总数")


# ==================== 因子计算 Schema ====================

class FactorCalculationRequest(BaseModel):
    """因子计算请求"""

    factor_id: int | None = Field(None, description="因子ID（None表示计算所有启用的因子）")
    codes: list[str] | None = Field(None, description="股票代码列表（None表示使用配置中的codes）")
    start_date: date | None = Field(None, description="开始日期（可选，参考日线数据处理逻辑：都不提供则使用最后一个交易日；部分提供时start_date默认2025-01-01）")
    end_date: date | None = Field(None, description="结束日期（可选，参考日线数据处理逻辑：都不提供则使用最后一个交易日；部分提供时end_date默认最后一个交易日）")


class FactorCalculationResponse(BaseModel):
    """因子计算响应"""

    success: bool = Field(..., description="是否成功")
    message: str = Field(..., description="消息")
    calculated_count: int = Field(0, description="计算的股票数量")
    failed_count: int = Field(0, description="失败的股票数量")
    details: list[dict[str, Any]] = Field(default_factory=list, description="详细信息")


# ==================== 因子结果 Schema ====================

class FactorResultItem(BaseModel):
    """因子结果项"""

    id: int | None
    trade_date: date
    factor_name: str | None = None
    factor_value: float | None

    class Config:
        from_attributes = True


class FactorResultResponse(BaseModel):
    """因子结果响应"""

    code: str = Field(..., description="股票代码")
    factor_name: str = Field(..., description="因子名称")
    items: list[FactorResultItem] = Field(..., description="因子结果列表")
    total: int = Field(..., description="总数")


class FactorResultQueryRequest(BaseModel):
    """因子结果查询请求"""

    code: str = Field(..., description="股票代码")
    factor_name: str | None = Field(None, description="因子名称（None表示查询所有）")
    trade_date: date | None = Field(None, description="交易日期")


# ==================== 量化因子查询 Schema ====================

class QuantFactorQueryRequest(BaseModel):
    """量化因子查询请求"""

    ts_code: str | None = Field(None, description="股票代码，模糊查询")
    start_date: date | None = Field(None, description="开始日期")
    end_date: date | None = Field(None, description="结束日期")
    filter_conditions: Any | None = Field(None, description="筛选条件，支持逻辑组合")
    skip: int = Field(0, description="跳过记录数")
    limit: int = Field(100, description="限制返回记录数")
    order_by: str | None = Field(None, description="排序字段")
    order: str | None = Field("desc", description="排序方式：asc或desc")


class QuantFactorQueryResponse(BaseModel):
    """量化因子查询响应"""

    items: list[dict[str, Any]] = Field(..., description="因子数据列表")
    total: int = Field(..., description="总记录数")
